File: src/14_longest_common_prefix/main.py
#!usr/bin/python

#
# \file
# \brief [Longest Common Prefix](https://leetcode.com/problems/longest-common-prefix/)
#
# \author Viachaslau Kukhtsiuk (https://github.com/Wenslauv)
#
# \details
# <h4>Problem</h4>
# Write a function to find the longest common prefix string amongst an array of strings.
#
# If there is no common prefix, return an empty string "".
#
#
# <h4>Constrains</h4>
#       1 <= strs.length <= 200
#       0 <= strs[i].length <= 200
#       strs[i] consists of only lowercase English letters.
#


from typing import List


def longestCommonPrefix(strs: List[str]) -> str:
    strs.sort()
    first, last = strs[0], strs[ -1 ]

    commonLength = min( len(first), len(last) )

    prefix = "";
    for i in range( commonLength ):
        if (first[i] == last[i]):
            prefix += first[i]
        else:
            return prefix

    return prefix

    
class Testcase:
    def test1():
        assert longestCommonPrefix( ["flower","flow","flight"] ) == "fl"


    def test2():
         assert longestCommonPrefix( ["dog","racecar","car"] ) == ""

    
    def test3():
         assert longestCommonPrefix( ["ab","a"] ) == "a"


    
    def runTests():
        print("Running predefined tests...")

        Testcase.test1()
        Testcase.test2()
        Testcase.test3()

        print("All tests passed successfully")





if __name__ == "__main__":
    Testcase.runTests()
    

File: src/14_longest_common_prefix/main.cpp
/**
 *
 * \file
 * \brief [Longest Common Prefix](https://leetcode.com/problems/longest-common-prefix/)
 *
 * \author Viachaslau Kukhtsiuk (https://github.com/Wenslauv)
 *
 * \details
 * <h4>Problem</h4>
 * Write a function to find the longest common prefix string amongst an array of strings.
 *
 * If there is no common prefix, return an empty string "".
 *
 *
 * <h4>Constrains</h4>
 *      1 <= strs.length <= 200
 *      0 <= strs[i].length <= 200
 *      strs[i] consists of only lowercase English letters.
 */


#include <cassert>
#include <iostream>
#include <string>
#include <vector>


namespace leetcode {

    std::string longestCommonPrefix(std::vector<std::string>& strs) {
        
        if (strs.size() == 1)
            return strs.front();

        int size            = static_cast<int>( strs.size() );
        std::string prefix  = strs.front();

        for (int i = 1; i < size; ++i) {
            auto& current = strs[i];

            if (current.size() < prefix.size()) {
                prefix.resize( current.size() );
            }

            int prefixSize = static_cast<int>( prefix.size() );
            for (int j = 0; j < prefixSize; ++j) {

                if (prefix[j] != current[j]) {
                    prefix.resize( j );
                    break;
                }

            }
        }

        return prefix;
    }


    namespace testing {

        void test1(void) {
            std::vector<std::string> strs = { "flower","flow","flight" };
            assert( longestCommonPrefix(strs) == "fl" );
        }

        void test2() {
            std::vector<std::string> strs = { "dog","racecar","car" };
            assert( longestCommonPrefix(strs) == "" );
        }

        void test3() {
            std::vector<std::string> strs = { "ab","a" };
            assert( longestCommonPrefix(strs) == "a" );
        }

        void runTests() {
            std::cout << "Running predefined tests..." << std::endl;
            
            test1();
            test2();
            test3();

            std::cout << "All tests passed successfully" << std::endl;
        }
    }   // namespace testing

}   // namespace leetcode



int main(int argc, const char * argv[]) {
    leetcode::testing::runTests();
    return 0;
}
